frontend/app/controllers/spree/store_controller.rb

```diff
@@ -26,6 +26,11 @@ def apply_coupon_code
         handler = PromotionHandler::Coupon.new(@order).apply
 
         if handler.error.present?
+          # If there are errors applying the coupon code we need to render the
+          # edit view and stop the request execution. Checkout steps need a
+          # setup before rendering the edit view or they will miss some ivars.
+          send(:setup_for_current_state) if respond_to?(:setup_for_current_state, true)
```

Comment:
maybe an alternative approach could be redirecting to the right checkout step?

Reply:
This is pointing to a larger problem in this bit of code. I think it might be best to apply some refactoring to here in order to avoid the `respond_to?` and `send`.

Some suggestions of things I would try:

1. Extract `apply_coupon_code` to a common base class (which extends `StoreController`) and modify the checkout controller and orders controller to inherit from that.
2. Extract the error handling to a method, allowing it to be overridden to handle the error condition and call `super` to finish the job.